source/MRViewer/MRSurfaceContoursWidget.cpp

```diff
@@ -0,0 +1,421 @@
+        obj_{ obj },
+        point_{ point },
+        index_{ index }
+    {};
+
+    virtual std::string name() const override
+    {
+        return "Move Point";
+    }
+    virtual void action( Type ) override
+    {
+        if ( !plugin_.isPickerActive_ )
+            return;
+
+        plugin_.pickedPoints_[obj_][index_]->updateCurrentPosition( point_ );
+        plugin_.activeIndex_ = index_;
+        plugin_.activeObject_ = obj_;
+        plugin_.onPointMoveFinish_( obj_ );
+    }
+
+    [[nodiscard]] virtual size_t heapBytes() const override
+    {
+        return 0; //this undo action will be deleted in plugin disable
+    }
+private:
+    SurfaceContoursWidget& plugin_;
+    const std::shared_ptr<MR::ObjectMeshHolder> obj_;
+    MeshTriPoint point_;
+    int index_;
+};
+
+void SurfaceContoursWidget::enable( bool isEnaled )
+{
+    isPickerActive_ = isEnaled;
+    if ( !isPickerActive_ )
+        pickedPoints_.clear();
+}
+
+std::shared_ptr<SurfacePointWidget> SurfaceContoursWidget::createPickWidget_( const std::shared_ptr<MR::ObjectMeshHolder> obj_, const MeshTriPoint& pt )
+{
+    auto newPoint = std::make_shared<SurfacePointWidget>();
+    newPoint->setAutoHover( false );
+    auto objMesh = std::dynamic_pointer_cast< MR::ObjectMesh > ( obj_ );
+    if ( objMesh )
+        newPoint->create( objMesh, pt );
+    else
+    {
+        assert( 1 > 2 ); // due to surfacePointWidget unable to work with non ObjectMesh objects.
+        return {};
+    }
+    std::weak_ptr<SurfacePointWidget> curentPoint = newPoint;
+    newPoint->setStartMoveCallback( [this, obj_, curentPoint] ( const MeshTriPoint& point )
+    {
+        const bool closedPath = pickedPoints_.size() > 1 && pickedPoints_[obj_][0]->getCurrentPosition() == pickedPoints_[obj_].back()->getCurrentPosition();
+        if ( closedPath && curentPoint.lock() == pickedPoints_[obj_][0] )
+        {
+            SCOPED_HISTORY( "Change Point" );
+            AppendHistory<ChangePointActionPickerPoint>( *this, obj_, point, activeIndex_ );
+            AppendHistory<ChangePointActionPickerPoint>( *this, obj_, point, int( pickedPoints_[obj_].size() ) - 1 );
+            moveClosedPoint_ = true;
+        }
+        else if ( !closedPath || curentPoint.lock() != pickedPoints_[obj_].back() )
+        {
+            AppendHistory<ChangePointActionPickerPoint>( *this, obj_, point, activeIndex_ );
+        }
+        activeChange_ = true;
+        onPointMove_( obj_ );
+
+    } );
+    newPoint->setEndMoveCallback( [this, obj_, curentPoint] ( const MeshTriPoint& point )
+    {
+        if ( moveClosedPoint_ && curentPoint.lock() == pickedPoints_[obj_][0] )
+        {
+            pickedPoints_[obj_].back()->updateCurrentPosition( point );
+        }
+        activeChange_ = false;
+        onPointMoveFinish_( obj_ );
+    } );
+
+    return newPoint;
+}
+
+bool SurfaceContoursWidget::isClosedCountour( const std::shared_ptr<ObjectMeshHolder> obj )
+{
+    return pickedPoints_[obj].size() > 1
```

Comment:
what if obj is not present in map?

Reply:
firs filter will return false. 
but we add pickedPoints_.count( obj ) > 0